torch/csrc/jit/codegen/cuda/fusion_executor.cpp

```diff
@@ -0,0 +1,589 @@
+
+std::string FusionExecutor::getKernel() {
+  // generating cuda code;
+  std::string code = std::string("namespace ") + FusionExecutor::Namespace() +
+      " {\n" + KernelUtils() + GPULower(&fusion_).getKernel(KernelName()) +
+      "}\n";
+
+  const char* debug_env = getenv("PYTORCH_CUDA_FUSER_DEBUG");
+  if (debug_env && atoi(debug_env)) {
+    std::cout << "\n==== codegen output for kernel: " << KernelName()
+              << " ====" << std::endl
+              << code << std::endl
+              << "====================================" << std::endl;
+  }
+
+  return code;
+}
+
+// TODO: add options for constraints, if we want to enforce a parallelization
+// strategy because we did something like split on a symbolic size.
+LaunchParams FusionExecutor::computeLaunchParams(
+    const at::ArrayRef<IValue>& aten_inputs) {
+  TORCH_INTERNAL_ASSERT(
+      fusion_.inputs().size() == aten_inputs.size(),
+      "Something went wrong configuring launch. Inputs no longer match.");
+
+  auto fusion_inputs = fusion_.inputs();
+  EvaluationContext ec(&fusion_);
+
+  // This should probably move to EvaluationContext as we may want to bind input
+  // values frequently. Bind fusion input values to runtime values.
+  for (size_t i = 0; i < fusion_.inputs().size(); i++) {
+    if (fusion_.inputs()[i]->getValType() == ValType::TensorView) {
+      TensorView* cg_tensor = fusion_.inputs()[i]->as<TensorView>();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_inputs[i].isTensor(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      auto aten_tensor = aten_inputs[i].toTensor();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_tensor.ndimension() == cg_tensor->getRootDomain().size(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      for (size_t dim = 0; dim < cg_tensor->getRootDomain().size(); dim++) {
+        ec.bind(
+            cg_tensor->getRootDomain()[dim]->extent(),
+            aten_tensor.sizes()[dim]);
+      }
+    }
+  }
+
+  LaunchParams lp;
+
+  // Grab all used values and run through them
+  auto unordered_vals = DependencyCheck::getAllValsBetween(
+      {fusion_inputs.begin(), fusion_inputs.end()}, fusion_.outputs());
+
+  // Grab only the TensorViews
+  std::set<TensorView*> unordered_tvs;
+  for (auto val : unordered_vals) {
+    if (val->getValType().value() == ValType::TensorView) {
+      TensorView* tv = val->as<TensorView>();
+      unordered_tvs.emplace(tv);
+    }
+  }
+
+  // Infer bindings
+  for (auto tv : unordered_tvs) {
+    for (auto id : tv->domain()->domain()) {
+      if (id->isThread()) {
+        auto val = ExpressionEvaluator::evaluate(id->rawExtent(), &ec);
+        TORCH_INTERNAL_ASSERT(
+            val,
+            "Tried to evaluate, ",
+            id,
+            ", within the tensor ",
+            tv,
+            " to set launch bounds but could not.");
+        lp.bind(val.value(), id->parallel_method());
```

Comment:
I'm asking a future-ish question here.

If we are to support broadcast of size-1 dimension, does that mean we could have an IterDomain bind to thread dimension, which has size-1, while non-broadcasted tensor binding to the same thread dimension would have a different size.

Reply:
We're going to have to figure out how that will work. I don't have the logic in for parallelized broadcast right now. I think we'll just exclude size=1 dims from the inference.